client/backoff/backoff_test.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backoff
+
+import (
+	"testing"
+	"time"
+)
+
+func TestBackoff(t *testing.T) {
+	b := Backoff{
+		Min:    time.Duration(1),
+		Max:    time.Duration(100),
+		Factor: 2,
+	}
+	for _, test := range []struct {
+		b     Backoff
+		times int
+		want  time.Duration
+	}{
+		{b, 1, time.Duration(1)},
+		{b, 2, time.Duration(2)},
+		{b, 3, time.Duration(4)},
+		{b, 4, time.Duration(8)},
+		{b, 8, time.Duration(100)},
+	} {
+		test.b.Reset()
+		var got time.Duration
+		for i := 0; i < test.times; i++ {
+			got = test.b.Duration()
+		}
+		if got != test.want {
+			t.Errorf("Duration() %v times: %v, want %v", test.times, got, test.want)
+		}
+	}
+}
+
+func TestJitter(t *testing.T) {
+	b := Backoff{
+		Min:    time.Duration(1) * time.Second,
+		Max:    time.Duration(100) * time.Second,
+		Factor: 2,
+		Jitter: true,
+	}
+	for _, test := range []struct {
+		b     Backoff
+		times int
+		min   time.Duration
+		max   time.Duration
+	}{
+		{b, 1, time.Duration(1) * time.Second, time.Duration(2) * time.Second},
+		{b, 2, time.Duration(2) * time.Second, time.Duration(4) * time.Second},
+		{b, 3, time.Duration(4) * time.Second, time.Duration(8) * time.Second},
+		{b, 4, time.Duration(8) * time.Second, time.Duration(16) * time.Second},
+		{b, 8, time.Duration(100) * time.Second, time.Duration(200) * time.Second},
```

Comment:
I thought about capping Jitter at Max, but I didn't want the max to cap things at a static value when jitter is on lets clients all clients wait the same amount of time. 
The current algorithm uses a geometrically increasing delay [A, B, C, D] = [c*a^0, c*a^1, c^a*2..]. Jitter selects a value from next higher interval. To keep max constant, jitter would have to select from the next lower interval, which would cause problems for x=0.